### PR TITLE
chore: Refactor notifications handling (AR-2184) (AR-2078)

### PR DESCRIPTION
**Pull request by borichellow**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-2184" title="AR-2184" target="_blank">BugAR-2184</a>  Notifications do not come
  </td></table>
  <br />
 

## Fixes [Bug-2184](https://wearezeta.atlassian.net/browse/AR-2184) , [Bug-2078](https://wearezeta.atlassian.net/browse/AR-2078)

# What's new in this PR?

Changed the flow of getting and displaying the notifications. 

Previously: `GetNotificationsUseCase` supposed to emit the whole list of notifications that should be displayed. If some notifications are not there -> it should be removed from the Top-bar. 
Such implementations have an obvious limitation: we should remember the previous notifications list, to compare it.

From now: `GetNotificationsUseCase` emits only the list of notifications that should be added to already displayed (just like current iOS app works), and `MessageNotificationManager` handle updating and adding the notifications.

This makes the whole notifications flow more intuitive (IMO) + we don't need a lot of "DismissReceivers" etc.

----

### Needs to be realized with: https://github.com/wireapp/kalium/pull/800

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [1](https://10.10.124.134/job/wire-android-reloaded/job/PR-867/1/) **failed**.

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-867/2/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(4184031)(4184031)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(4184031)(4184031)-dev-release.apk)

**Comment by AndroidBob** [comment]
Build [3](https://10.10.124.134/job/wire-android-reloaded/job/PR-867/3/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(4185137)(4185137)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(4185137)(4185137)-dev-release.apk)

File: app/src/main/kotlin/com/wire/android/notification/MessageNotificationManager.kt [hunk 1]
```diff
@@ -26,41 +30,40 @@ import javax.inject.Singleton
...
+
+        if (notificationsLeft <= 1) {
+            // only SummaryNotification left
+            notificationManager.cancel(NotificationConstants.MESSAGE_SUMMARY_ID)
+        }
+    }
+
+    fun hideAllNotifications() {
+        // removing groupSummary removes all the notifications in a group
+        notificationManager.cancel(NotificationConstants.MESSAGE_SUMMARY_ID)
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/notification/MessageNotificationManager.kt:66]: Is this `MESSAGE_SUMMARY_ID` grouping ALL kind of notifications or only text related? Just asking out of ignorance here.

Comment (borichellow) [reply app/src/main/kotlin/com/wire/android/notification/MessageNotificationManager.kt:66]: it groups everything except the calls, calls are in the separate channel

File: app/src/main/kotlin/com/wire/android/notification/MessageNotificationManager.kt [hunk 2]
```diff
@@ -72,41 +75,49 @@ class MessageNotificationManager @Inject constructor(private val context: Contex
...
-        .setDeleteIntent(dismissSummaryPendingIntent(context, userId))
         .build()
 
-    private fun getConversationNotification(conversation: NotificationConversation, userId: String?) =
-        NotificationCompat.Builder(context, NotificationConstants.MESSAGE_CHANNEL_ID).apply {
+    /**
+     * Create or update existed [Notification]
+     * @return [Notification] for the conversation with all the messages in it (including the messages that been there before)
+     * OR null if there is no new messages in conversation and no need to update the existed notification.
+     */
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/notification/MessageNotificationManager.kt:111]: Small suggestion to improve the provided docs
```suggestion
    /**
     * Creates or updates existing [Notification]
     * @param [conversation] the notification object containing all the messages for the given conversation
     * @param [userId] the id of the user receiving the notifications
     * @param [activeNotifications] a list with the notifications that should be displayed to the user
     * @return [Notification] for the conversation with all the messages in it (including previous messages as well)
     * OR null if there is no new messages in conversation and no need to update the existed notification.
     */
```

**Comment by AndroidBob** [comment]
Build [4](https://10.10.124.134/job/wire-android-reloaded/job/PR-867/4/) **failed**.

File: app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -85,10 +85,7 @@
         </provider>
 
         <receiver android:name=".notification.broadcastreceivers.NotificationReplyReceiver"/>
-        <receiver android:name=".notification.broadcastreceivers.MessageNotificationDismissReceiver"/>
         <receiver android:name=".notification.broadcastreceivers.CallNotificationDismissReceiver"/>
-        <receiver android:name=".notification.broadcastreceivers.ConnectionRequestNotificationDismissReceiver"/>
-        <receiver android:name=".notification.broadcastreceivers.SummaryNotificationDismissReceiver"/>
```
[the remark this answers is not in the document]
Comment (borichellow) [reply app/src/main/AndroidManifest.xml:91]: as I described in PR description, from now on: `GetNotificationsUseCase` emits only the list of notifications that should be added to the already displayed (just like the current iOS app works).
With such implementation when `GetNotificationsUseCase` we just add notifications, mark it as "notified" (by `MarkMessagesAsNotifiedUseCase` and `MarkConnectionRequestAsNotifiedUseCase`) and don't care what happens later :)

**Comment by AndroidBob** [comment]
Build [5](https://10.10.124.134/job/wire-android-reloaded/job/PR-867/5/) **failed**.

**Review by vitorhugods — APPROVED** [review]
✅

**Comment by AndroidBob** [comment]
Build [6](https://10.10.124.134/job/wire-android-reloaded/job/PR-867/6/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(4195662)(4195662)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(4195662)(4195662)-dev-release.apk)

**Pull request merged**